bulldozer/signals.go

```diff
@@ -21,109 +21,389 @@ import (
 	"strings"
 
 	"github.com/palantir/bulldozer/pull"
+	"github.com/pkg/errors"
 	"github.com/rs/zerolog"
 )
 
+type Signal []string
+
 type Signals struct {
-	Labels            []string `yaml:"labels"`
-	CommentSubstrings []string `yaml:"comment_substrings"`
-	Comments          []string `yaml:"comments"`
-	PRBodySubstrings  []string `yaml:"pr_body_substrings"`
-	Branches          []string `yaml:"branches"`
-	BranchPatterns    []string `yaml:"branch_patterns"`
+	Labels            Signal `yaml:"labels"`
+	CommentSubstrings Signal `yaml:"comment_substrings"`
+	Comments          Signal `yaml:"comments"`
+	PRBodySubstrings  Signal `yaml:"pr_body_substrings"`
+	Branches          Signal `yaml:"branches"`
+	BranchPatterns    Signal `yaml:"branch_patterns"`
+	MaxCommits        int `yaml:"max_commits"`
 }
 
-func (s *Signals) Enabled() bool {
+type SignalsMatches struct {
+	Signals
+}
+
+func (s *Signals) size() int {
 	size := 0
 	size += len(s.Labels)
 	size += len(s.CommentSubstrings)
 	size += len(s.Comments)
 	size += len(s.PRBodySubstrings)
 	size += len(s.Branches)
 	size += len(s.BranchPatterns)
-	return size > 0
+	if s.MaxCommits > 0 {
+		size += 1 
+	}
+	return size
 }
 
-// Matches returns true if the pull request meets one or more signals. It also
+func (s *Signals) Enabled() bool {
+	return s.size() > 0
+}
+
+// MatchesAll returns true if the pull request matches ALL of the signals. It also
```

Comment:
I initially thought `MatchesAll` meant that the pull request had to match the label condition and the comment condition and etc. But it looks like this also changes the meaning of the individual conditions so that "matching the label condition" now means that all labels in the list must exist.

I'm not sure that behavior is desirable or intuitive, but I could be convinced otherwise. How were you thinking about the match any vs match all behavior?

Reply:
I added this new matching logic to accommodate the new `max_commits` condition. Specifying a max number of commits in a PR was requested as a condition which would be used in conjunction with other trigger criteria. Since the merge methods are defined in a list which can be used to create "OR" logic, it felt logical to just "AND" all the triggers.

You are right this also changes the way triggering each signal works. I didn't consider that as a major issue when I was developing this, but I agree it's confusing. I'll go ahead and revert that.